File: SlidingWIndow/904FruitIntoBaskets.cpp
/**
 Solution:
 Sliding Window
*/
class Solution {
public:
    int totalFruit(vector<int>& tree) {
        unordered_map<int, int> m;
        int rev = 0;
        for(int i=0, j=0; j<tree.size(); j++){
            m[tree[j]]++;
            if(m.size()>2){
                m[tree[i]]--;
                if(m[tree[i]]==0) m.erase(tree[i]);
                i++;
            }
            rev = max(rev, j-i+1);
        }
        return rev;
    }
};

File: SlidingWIndow/904fruitIntoBaskets.py
class Solution(object):
    def totalFruit(self, tree):
        """
        :type tree: List[int]
        :rtype: int
        """
        dic = collections.Counter()
        start, res =0,0
        for i in range(len(tree)):
            dic[tree[i]] += 1
            while len(dic) > 2:
                dic[tree[start]] -= 1
                if dic[tree[start]] == 0:
                    del dic[tree[start]]
                start += 1
            res = max(res,i-start+1)
        return res
    
    